rust/connlib/shared/src/interface/linux.rs

```diff
@@ -0,0 +1,296 @@
+//! Virtual network interface
+
+use crate::{
+    callbacks::{Cidrv4, Cidrv6},
+    linux::{etc_resolv_conf, DnsControlMethod},
+};
+use anyhow::{anyhow, bail, Context as _, Result};
+use futures::TryStreamExt;
+use ip_network::{IpNetwork, Ipv4Network, Ipv6Network};
+use netlink_packet_route::route::{RouteProtocol, RouteScope};
+use netlink_packet_route::rule::RuleAction;
+use rtnetlink::{new_connection, Error::NetlinkError, Handle, RouteAddRequest, RuleAddRequest};
+use std::{
+    collections::HashSet,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+};
+
+// TODO: De-dupe before merging
+const FIREZONE_MARK: u32 = 0xfd002021;
+const IFACE_NAME: &str = "tun-firezone";
+const DEFAULT_MTU: u32 = 1280;
+const FILE_ALREADY_EXISTS: i32 = -17;
+const FIREZONE_TABLE: u32 = 0x2021_fd00;
+
+/// For lack of a better name
+pub struct InterfaceManager {
+    // This gets lazy-initialized when the interface is first configured
+    connection: Option<Connection>,
+    dns_control_method: Option<DnsControlMethod>,
+    routes: HashSet<IpNetwork>,
+}
+
+struct Connection {
+    handle: Handle,
+    task: tokio::task::JoinHandle<()>,
+}
+
+impl Drop for InterfaceManager {
+    fn drop(&mut self) {
+        if let Some(connection) = self.connection.take() {
+            connection.task.abort();
+        }
+        tracing::debug!("Reverting DNS control...");
+        if let Some(DnsControlMethod::EtcResolvConf) = self.dns_control_method {
+            // TODO: Check that nobody else modified the file while we were running.
+            etc_resolv_conf::revert().ok();
+        }
+    }
+}
+
+impl Default for InterfaceManager {
+    fn default() -> Self {
+        let dns_control_method = crate::linux::get_dns_control_from_env();
+        tracing::info!(?dns_control_method);
+
+        Self {
+            connection: None,
+            dns_control_method,
+            routes: Default::default(),
+        }
+    }
+}
+
+impl InterfaceManager {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub async fn on_set_interface_config(&mut self, ipv4: Ipv4Addr, ipv6: Ipv6Addr) -> Result<()> {
```

Comment:
```suggestion
    pub async fn set_ips(&mut self, ipv4: Ipv4Addr, ipv6: Ipv6Addr) -> Result<()> {
```

I'd suggest to move away from the naming of the callback functions here and consider this component in isolation.